### ISSUE TITLE
Problem with required and dependencies

### ISSUE THREAD

**Issue by evga** [nicolaiarocci/cerberus#274]
Used Cerberus version 1.0.1 installed with pip
- I consulted these documentations:
  - [ Validation Rules] http://docs.python-cerberus.org/en/stable/validation-rules.html
#### Support request / Bug report

According to the docs, the following program should validate successfully since 
the `auth-key` field is required only when the `auth` field is `wep` or `wpa`.

``` python

import cerberus

schema = {
    'auth' : {
        'type':'string',
        'required':True,
        'allowed':['open','wep','wpa']
    },
    'auth-key' : {
        'type':'string',
        'required':True,
        'dependencies': {'auth': ['wep', 'wpa']}
    }
}

v = cerberus.Validator(schema)
print v.validate({'auth':'open'})
print v.errors
```

The output of the program is the following:

```
False
{'auth-key': ['required field']}
```

Is this a bug ?

**Comment by gtnx** [comment]
I don't think it's a bug. You need to be more precise when specifying the auth-key:

```python
schema = {
    'auth' : {
        'type':'string',
        'required':True,
        'allowed':['open','wep','wpa']
    },
    'auth-key' : {
        'type':'string',
        'oneof': [{
            'required':True,
        }, {
            'dependencies': {'auth': ['open']},
        }]
    }
}
```

It's like saying: "wether auth-key is defined or auth is 'open'".

**Comment by aguirrel** [comment]
@gtnx I think this is a bug or documentation is not ok, I have the same situation.

**Comment by drcraig** [comment]
I think I'm also hitting this issue, too. The [documentation](http://docs.python-cerberus.org/en/stable/validation-rules.html#required) says 

> Also, if dependencies are declared for the field, its required rule will only be validated if all dependencies are included with the document.

In @aguirrel's example, I interpret that to mean that auth-key will only be required if auth's value is 'open'. Is that the intended behavior or am I misunderstanding?

**Comment by aguirrel** [comment]
@drcraig, in @evga example, auth-key field should be required if value is 'wep' or 'wpa' (dependencies).

**Comment by aguirrel** [comment]
I'm doing a regression test and it seems that in version 0.9 @evga example works fine. It's broken from 0.9.1 and above.

**Comment by drcraig** [comment]
@aguirrel Right, my mistake. I was projecting my own situation where my dependency is only one particular value. Thanks for running the regression tests.

**Comment by funkyfuture** [comment]
i think there's a misunderstanding here. looking at the code, there's no interaction between the `dependencies` and the `required` rule.

> Also, if dependencies are declared for the field, its required rule will only be validated if all dependencies are included with the document.

means: if a `dependencies` constraint failed for a field, its (possible) `required` will even not be evaluated. fun fact: i don't see that in the code neither.

concerning your use-case: make use of `oneof`, but at a higher level than @gtnx suggested.

this issue relates to #229. there's :cake: for everyone who shares her/his solutions.

**Comment by aguirrel** [comment]
@funkyfuture thanks a lot for clarifying the issue.  

What do you think about that posible interaction between dependencies and required?  Should it work as described in documentation or it's ok as it is working now?

If you ask me, I think it should work as ['exclude' validator](https://github.com/nicolaiarocci/cerberus/blob/master/cerberus/tests/test_validation.py#L1299), but i'm afraid to broke legacy code that depends on the current logic.  Should an 'include' validator be created to work similar to 'exclude' in 'required' use cases?

Could you reformulate @gtnx to be more explicit as you told us in your comment? I would like having positive validations ('dependencies': {'auth': ['wep', 'wpa']) instead of negative ones ('dependencies': {'auth': ['open']})

Thanks a lot!

**Comment by aguirrel** [comment]
@funkyfuture I have just sent a PR with a fix and a test case for this issue. Now I want to claim my :cake: :smile:

**Comment by funkyfuture** [comment]
```python
schemas = [
    {'auth': {'type': 'string', 'allowed': 'open'}},
    {'auth': {'type': 'string', 'allowed': ['wep', 'wpa']}
     'auth-key': {'type': 'string', 'required': True}}
]

validates = any((validator(doc, x) for x in schemas))
```

i'm not sure if the proposed change is really necessary, considering that it does change semantics in a way that existing code may validate circumstances that it wasn't intended to.

the cake is strictly for contributions regarding #229!

**Comment by aguirrel** [comment]
I think that @evga example is pretty simple and semantics is just what is expected (in my opinion).  My deduction is: required depends on field should be there in the first place.

* Why should be a field be required if depends on some field that is not is present or has a valid data?
* Why 'excludes' works and 'dependecies' no, if both are about the existence of the required field?

ok with the cake :pensive:

**Comment by drcraig** [comment]
I guess I'm either still confused and doing something wrong, or there is a bug.

```
schema = {
    'mode': {
        'allowed': ['auto', 'manual'],
        'required': True
    },
    'needed-for-auto': {
        'type': 'list',
        'dependencies': {'mode': 'auto'},
    }
}
```
My understanding is that this should mean that needed-for-auto would be required when mode is set to auto and not required otherwise.

```
# Should validate and does
document = {
    'mode': 'auto',
    'needed-for-auto': ['foo']
}      

# Should validate and does
document = {
    'mode': 'manual'
}

# Should not validate because it's missing needed-for-auto, but does validate
document = {
    'mode': 'auto'
}
```
Am I getting the dependency relationship backwards? I've tried various combinations of anyoff and allof like @gtnx showed, but none of them seem to work.

**Comment by drcraig** [comment]
Flipping the dependency seems to work.

```
schema = {
    'mode': {
        'oneof': [
            {'allowed': ['auto'], 'dependencies': 'needed-for-auto'},
            {'allowed': ['manual']}
        ],
        'required': True
    },
    'needed-for-auto': {
        'type': 'list'
    }
}
```

It took some time to wrap my head around it and get it right. Probably something that would be a good example in #229 .

**Comment by evga** [comment]
I'll add my two cents since I'm the one who opened the issue. I think that obvious things should just work, without complicated hacks that make the rules obscure and hard to understand. Before opening this issue i've read the docs many times and spent almost an hour trying to understand why it didn't work. In the end I realized that I was fighting against the library instead of using it to solve my problem, so I gave up and coded the validation logic by hand.

To avoid confusion I'll post here the relevant lines extracted from the documentation of the "required" validation option:

> Note
> String fields with empty values will still be validated, even when required is set to True. If you don’t want to accept empty values, see the empty rule. **Also, if dependencies are declared for the field, its required rule will only be validated if all dependencies are included with the document.**

I am no english expert.. but what I think when I see this phrase in the docs is that a required field is NOT required if the dependencies of that field are missing. In other words, I'm not required to tell you my wireless password if I'm going with open authentication.

I think that such a rule is actually very common in validation, so common that it should be pretty straightforward to implement and work out of the box without obscure hacks.

From what I see here there is either a false statement in the docs or a bug in the code.
If fixing the code is not possible due to backwards compatibility then fix the docs.

IMHO, adding more weird non-intuitive rules to make it work is not a good solution.

**Comment by aguirrel** [comment]
Hi guys, if someone need a solution for this, I have ended doing my own validator.  Main problem here is that if a field is not present in the doc, validations rules for this field are never called. That's why @gtnx 'oneof' works when field 'auth' is 'open' but gives True when 'auth' is 'wep' when should be False.  Indeed it works because it is never call.

```python
schema = {
    'auth' : {
        'type':'string',
        'required':True,
        'allowed':['open','wep','wpa']
    },
    'auth-key' : {
        'type':'string',
        'oneof': [{
            'required':True,
        }, {
            'dependencies': {'auth': ['open']},
        }]
    }
}

v=Validator()
v.schema = schema
v.validate({'auth': 'wep'}) # Gives True when should be False, and 'auth-key' should have required error
```

My solution is pretty simple, I check for fields in SCHEMA and get fields with 'requiredif' rule.

```python
class MyValidator(Validator):
    def validate_required_if(self, field, data, doc):
        for key, value in data['requiredif'].items():
            if isinstance(value, str) or isinstance(value, unicode):
                value = [value]
            if (key in doc and
                doc[key] in value and
                not field in doc):
                    self._error(field, "Required field")
                    return False
        return True

    def validate(self, doc):
        ret = super(ISOValidator, self).validate(doc)
        for key, data in self.schema.items():
            if 'requiredif' in data:
                ret = ret and self.validate_required_if(key, data, doc)
        return ret
```

**Comment by funkyfuture** [comment]
i had a look at the code history today and i found out that dependencies constraints once were actually taken into account when evaluating required. that was removed later (i didn't inquire that further), but the docs weren't changed accordingly. this is now included with #292.
